src/runners/rfq_maker_balance_cache_runner.ts

```diff
@@ -0,0 +1,169 @@
+    labelNames: ['workerId'],
+});
+
+process.on('uncaughtException', err => {
+    logger.error(err);
+    process.exit(1);
+});
+
+process.on('unhandledRejection', err => {
+    if (err) {
+        logger.error(err);
+    }
+});
+
+interface BalancesCallInput {
+    addresses: string[];
+    tokens: string[];
+}
+
+if (require.main === module) {
+    (async () => {
+        logger.info('running RFQ balance cache runner');
+        const ethersProvider = new ethers.providers.JsonRpcProvider(
+            defaultHttpServiceWithRateLimiterConfig.ethereumRpcUrl,
+        );
+        const connection = await getDBConnectionAsync();
+        const balanceCheckerContractInterface = await _getBalanceCheckerContractInterfaceAsync(ethersProvider);
+
+        await runRfqBalanceCheckerAsync(ethersProvider, connection, balanceCheckerContractInterface);
+    })().catch(error => logger.error(error.stack));
+}
+
+async function runRfqBalanceCheckerAsync(
+    ethersProvider: ethers.providers.JsonRpcProvider,
+    connection: Connection,
+    balanceCheckerContractInterface: ethers.Contract,
+): Promise<void> {
+    const workerId = _.uniqueId('rfqw_');
+    let lastBlockSeen = -1;
+    while (true) {
+        const newBlock = await ethersProvider.getBlockNumber();
+        if (lastBlockSeen < newBlock) {
+            lastBlockSeen = newBlock;
+            LATEST_BLOCK_PROCESSED_GAUGE.labels(workerId).set(lastBlockSeen);
+            logUtils.log(
+                {
+                    block: lastBlockSeen,
+                    workerId,
+                },
+                'Found new block',
+            );
+
+            const makerTokens = await _getMakerTokensAsync(connection);
+            const balancesCallInput = _splitValues(makerTokens);
+
+            const updateTime = new Date();
+            const erc20Balances = await _getErc20BalancesAsync(balanceCheckerContractInterface, balancesCallInput);
+
+            await _updateErc20BalancesAsync(balancesCallInput, erc20Balances, connection, updateTime);
+
+            await delay(DELAY_WHEN_NEW_BLOCK_FOUND);
+        } else {
+            await delay(DELAY_WHEN_NEW_BLOCK_NOT_FOUND);
+        }
+    }
+}
+
+// NOTE: this only returns a partial entity class, just token address and maker address
+// Cache the query results to reduce reads from the DB
+let MAKER_TOKEN_CACHE: ResultCache<MakerBalanceChainCacheEntity[]>;
+const _getMakerTokensAsync = async (connection: Connection) => {
+    if (!MAKER_TOKEN_CACHE) {
+        logUtils.log('updating cache');
+        MAKER_TOKEN_CACHE = createResultCache<any[]>(
+            () =>
+                connection
+                    .getRepository(MakerBalanceChainCacheEntity)
+                    .createQueryBuilder('maker_balance_chain_cache')
+                    .select(['maker_balance_chain_cache.tokenAddress', 'maker_balance_chain_cache.makerAddress'])
+                    .getMany(),
+            TEN_MINUTES_MS,
```

Comment:
This may be a problem, unfortunately.

In the cache validator component, we assume a cache entry is invalid if more than 5 minutes have passed since the entry was first seen and no balances exist. If we cache results for 10 minutes in your code, this may cause new address entries to be seen as stale. In [this code](https://github.com/0xProject/0x-api/pull/432/files#diff-fa95b01194bc9cd84720be5a299890448459cfb7c4b7faa452681c9f11f65907R151), `_cacheExpiryThresholdMs` will default to 5 seconds. The code is changeable of course, so let's just sync up and discuss a solution that will work with both of our codes.

However, I may suggest thinking if a cache is really needed? this query is fairly simple, the results should have very low cardinality (~100 rows max) and, unlike the API `swap` request when we verify maker balances using this table, there is no "added pressure" on the DB or on request latency if this PG query takes more time than expected. Would it probably be better to simply add a gauge (using Prometheus) and simply benchmark latency? if latency becomes a problem, we can fix the issue?

Reply:
reduce to 30s, keep tabs on read latency

Reply:
```typescript
const MAKER_CACHE_RESULT_COUNT = new Counter({
    name: 'maker_cache_result_count',
    help: 'Add me here',
    labelNames: [],
});
```
```